[MISC] Add Handwritten English Text Recognition Model

jnzw opened the pull request:
Hi, this PR is related to the [handwritten-english-recognition-0001 OMZ model](https://github.com/openvinotoolkit/open_model_zoo/tree/master/models/intel/handwritten-english-recognition-0001).
Could you please take a look?

Modifications:
* Add yaml model config file
* Add transformations for data augmentation and normalization
* Add vocab json file for [GNHK dataset](https://goodnotes.com/gnhk/) 
* Add custom CNN backbone
* Updated accuracy calculation (we used a different metric, character error rate instead of word error rate)
* Add dropout option to LSTM
* Update options in the model exporter
* About the `whitespace` updates: we trained a text line recognition model instead of word image recognition model, so spaces are in the ground truth sentences, hence the related modifications
* I have the GNHK dataset in LMDB format, the saved checkpoint, and onnx files ready. If needed, please let me know where to upload them

Thanks.

jnzw committed:
Add model

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Update exporter and configs

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
update configs

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Update README

Signed-off-by: Junze Wu <junze.wu@intel.com>

morkovka1337 commented on misc/pytorch_toolkit/text_recognition/text_recognition/data/utils.py:
```diff
@@ -64,6 +65,64 @@ def __call__(self, imgs):
+            A.Perspective(scale=0.05, fit_output=True, pad_val=255, p=0.5),
+            A.RandomBrightnessContrast(p=0.5),
+            A.GaussNoise(p=0.5)
+        ])
```
I would generalize this class into a class representing albumentations transforms. Now the list of transforms is specific for a concrete model. It can be generalized in the following way (did not test if it works correctly):
config file: 
```
train_transforms_list:
- name: "TransformAlbumentations"
  transforms_list:
  - type="RandomScale"
    scale_limit=0.2
    p=0.5
  - type="Affine"
    shear=[-0.5,0.5] 
    fit_output=true
    p=0.5
    cval=255
# and other transforms
```
This class `__init__`:
```
self.transform_list = []
for transform in transforms_list:
    self.transform_list.append(getattr(A, transform.pop('type'))(**transform)
self.transform_list = A.Compose(self.transform_list)
```

This will allow to use any transforms available in albumentations and set them directly in the config file without code changes.

morkovka1337 commented on misc/pytorch_toolkit/text_recognition/text_recognition/models/backbones/cnn.py:
```diff
@@ -0,0 +1,52 @@
+import torch.nn as nn
+
+class ConvNetBackbone(nn.Module):
```
Consider changing class name into something more specific. `ConvNetBackbone` is too general, maybe something like `HETRBackbone`.

morkovka1337 commented on misc/pytorch_toolkit/text_recognition/text_recognition/utils/trainer.py:
```diff
@@ -340,22 +367,34 @@ def validate(self, use_gt_token=True):
+                                                                          ignore_end_token=self.config.get('use_ctc'),
+                                                                          whitespace=self.whitespace)
+                            for param in self.config.get('datasets')['validate']:
+                                case_sensitive = param.get('case_sensitive', False)
```
```suggestion
                            case_sensitive = getattr(loader.dataset, "case_sensitive", False)
```
In current implementation `case_sensitive` setting is got from the last dataset in the list. Plus, such implementation allows to perform case-sensitive testing for specific dataset (e.g., when you are evaluating on two datasets with different setting `case_sensitive`).

morkovka1337 reviewed (changes requested):
Junze, thank you for your hard work! In addition to code comments, please:

- [ ] Add small amount of sample images (10-20 will be enough) to perform tests into the [data](https://github.com/openvinotoolkit/training_extensions/tree/master/data) folder. 
- [ ] Add tests for the new model to make sure it works correctly. 
- [ ] Add a file containing reference metric the new model achieves on the __added toy dataset__. You can see scene text recognition 14, 15, 16 models as a reference. 
- [ ] Add documentation into readme file about new backbone and supported tasks.

jnzw committed:
Update evaluator

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Add unit tests

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Add export test

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Fix case_sensitive

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Rename ConvNetBackbone

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Update README

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Fix test err

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Update albumentations transformation

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Update comment

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Update backbone name

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
remove extra empty line

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Remove sample images due to legal concern

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Add line space between classes

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Update config name

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Update config name

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Add small toydataset

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw commented:
@morkovka1337 Ready for another round of review now.

- [x] Added 10 images created by hand to `data/handwritten_text_recognition/`

- [x] Added test cases (train, eval, export) , all pass locally
- [x] Added metric for the toy dataset to `tests/expected_outputs/handwritten_recognition/hetr.json`
- [x] Updated README
- [x] Upload model to storage

jnzw committed:
Add ground truth file

Signed-off-by: Junze Wu <junze.wu@intel.com>

morkovka1337 commented on misc/pytorch_toolkit/text_recognition/text_recognition/models/model.py:
```diff
@@ -20,6 +20,7 @@
 import torch.nn as nn
 
 from .backbones.resnet import (CustomResNetLikeBackbone, ResNetLikeBackbone)
+from .backbones.cnn import HETRBackbone
```
Please, change filename from `cnn` to `hetr` also, and rename class name in the backbones mapping and in the config (from `cnn` to `hetr`)

morkovka1337 commented on data/handwritten_text_recognition/create_lmdb_dataset.py:
```diff
@@ -0,0 +1,87 @@
+""" a modified version of CRNN torch repository https://github.com/bgshih/crnn/blob/master/tool/create_dataset.py """
```
Intel header need to be added here (in addition to the note about file origin).
```suggestion
""" 
 Copyright (c) 2021 Intel Corporation
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
      http://www.apache.org/licenses/LICENSE-2.0
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 
 
a modified version of CRNN torch repository https://github.com/bgshih/crnn/blob/master/tool/create_dataset.py 
"""
```

jnzw committed:
Rename backbone config

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Update data/handwritten_text_recognition/create_lmdb_dataset.py

Co-authored-by: Vladimir Loginov <vladimir.loginov@intel.com>

jnzw committed:
Update correct link to pth

Signed-off-by: Junze Wu <junze.wu@intel.com>

morkovka1337 commented:
run tests

morkovka1337 commented on misc/pytorch_toolkit/text_recognition/README.md:
```diff
@@ -171,6 +173,7 @@ Here you can find which models are currently supported. For accuracy metrics and
 | text-recognition-0014        | Alphanumeric Scene Text Recognition | None           | ResNeXt-50 st 2  | CTC LSTM Encoder-Decoder | [link](https://github.com/openvinotoolkit/open_model_zoo/tree/master/models/intel/text-recognition-0014)                            | [link](https://storage.openvinotoolkit.org/repositories/openvino_training_extensions/models/text_recognition/text_recognition_0014. ...
 | text-recognition-0015        | Alphanumeric Scene Text Recognition | None           | ResNeXt-101      | 2d attention             | [link](https://github.com/openvinotoolkit/open_model_zoo/tree/master/models/intel/text-recognition-0015)                            | [link](https://storage.openvinotoolkit.org/repositories/openvino_training_extensions/models/text_recognition/text_recognition_0015. ...
 | text-recognition-0016 ([YATR](https://arxiv.org/abs/2107.13938)) | Alphanumeric Scene Text Recognition | TPS            | ResNeXt-101 st 3 | 2d attention             | [link](https://github.com/openvinotoolkit/open_model_zoo/tree/develop/models/intel/text-recognition-0016)                                                                                                                             ...
```
```suggestion
| text-recognition-0016 ([YATR](https://arxiv.org/abs/2107.13938)) | Alphanumeric Scene Text Recognition | TPS            | ResNeXt-101 st 3 | 2d attention             | [link](https://github.com/openvinotoolkit/open_model_zoo/tree/master/models/intel/text-recognition-0016)                                                                                                                                 | [link](https://storage.openvinotoolkit.org/repositories/openvino_training_extensions/models/text_recognition/text_recognition_0016.pth) | [link](configs/config_0016.yml)                    |
```

morkovka1337 reviewed (changes requested):
Open Model Zoo repo was updated, links with `develop` branch are no longer alive. Please, update corresponding link in the 196 line of the readme file, I'm unable to do this via github interface.

jnzw committed:
Update misc/pytorch_toolkit/text_recognition/README.md

fix broken link

Co-authored-by: Vladimir Loginov <vladimir.loginov@intel.com>

morkovka1337 commented:
run tests

jnzw committed:
fix broken links

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw committed:
Merge branch 'hetr' of https://github.com/jnzw/training_extensions into hetr

jnzw committed:
Fix linting error

Signed-off-by: Junze Wu <junze.wu@intel.com>

jnzw commented:
Hi, I looked at the Jenkins console output and there were 2 documentation broken links in text_recognition/README.md and some lint errors in my code, which all should have been fixed now (please see the recent commits). There are some errors in ote_tests_runner that seems unrelated to this PR (one for image classification and one for object detection).

morkovka1337 commented:
run tests

morkovka1337 commented:
Seems like this is ready to be merged. We will fix broken tests for the classification and detection via separate PR. @Ilya-Krylov, please, merge.
